### PR TITLE
OTC-354: Premiums wrongly assigned to policies during enrolment

### PR DESCRIPTION
**Pull request by malinowskikam**
Changes:
- Policies can have multiple premiums
- Premiums are now assigned based on IDs

[https://openimis.atlassian.net/browse/OTC-354](https://openimis.atlassian.net/browse/OTC-354)

### CODE REVIEW

File: app/src/main/java/org/openimis/imispolicies/ClientAndroidInterface.java [hunk 1]
```diff
@@ -3638,15 +3638,21 @@ private int Enrol(int oFamilyId, int oInsureeId, int oPolicyId, int oPremiumId,
...
                         // Policy + premium
-                        JSONArray tempPoliciesArray = new JSONArray();
 
                         for (int j = 0; j < policiesArray.length(); j++) {
-                            tempPoliciesArray = policiesArray;
-                            JSONObject premiumObj = premiumsArray.getJSONObject(j);
-                            tempPoliciesArray.getJSONObject(j).put("premium", premiumObj);
+                            JSONObject policyObject = policiesArray.getJSONObject(j);
+                            JSONArray policyPremiums = new JSONArray();
+                            String policyId = policyObject.getString("PolicyId");
```
Comment (kmadej) [inline app/src/main/java/org/openimis/imispolicies/ClientAndroidInterface.java:3645]: there is no need to create this local var

File: app/src/main/java/org/openimis/imispolicies/ClientAndroidInterface.java [hunk 2]
```diff
@@ -3638,15 +3638,21 @@ private int Enrol(int oFamilyId, int oInsureeId, int oPolicyId, int oPremiumId,
...
+                            for(int k = 0; k < premiumsArray.length(); k++) {
+                                JSONObject premiumObject = premiumsArray.getJSONObject(k);
+                                if(policyId.equals(premiumObject.getString("PolicyId"))) {
```
Comment (kmadej) [inline app/src/main/java/org/openimis/imispolicies/ClientAndroidInterface.java:3648]: please use string utils equals to omit npe problem

**Review by kmadej — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/openimis/imispolicies/ClientAndroidInterface.java (hunk 2 above)
Comment (malinowskikam) [reply app/src/main/java/org/openimis/imispolicies/ClientAndroidInterface.java]: I added StringUtils to the project, waiting for approval to use Apache licensed libraries.

File: app/src/main/java/org/openimis/imispolicies/ClientAndroidInterface.java [hunk 3]
```diff
@@ -3638,15 +3639,20 @@ private int Enrol(int oFamilyId, int oInsureeId, int oPolicyId, int oPremiumId,
...
 
                         for (int j = 0; j < policiesArray.length(); j++) {
-                            tempPoliciesArray = policiesArray;
-                            JSONObject premiumObj = premiumsArray.getJSONObject(j);
-                            tempPoliciesArray.getJSONObject(j).put("premium", premiumObj);
+                            JSONArray policyPremiums = new JSONArray();
+                            String policyId = policiesArray.getJSONObject(j).getString("PolicyId");
+                            for(int k = 0; k < premiumsArray.length(); k++) {
+                                JSONObject premiumObject = premiumsArray.getJSONObject(k);
+                                if(StringUtils.equals(policyId,premiumObject.getString("PolicyId"))) {
```
Comment (kmadej) [inline app/src/main/java/org/openimis/imispolicies/ClientAndroidInterface.java:3648]: whitespace after ",". Please move PolicyID string to constant

Comment (kmadej) [reply app/src/main/java/org/openimis/imispolicies/ClientAndroidInterface.java:3648]: whitespace after "if"

File: app/src/main/java/org/openimis/imispolicies/ClientAndroidInterface.java [hunk 4]
```diff
@@ -3638,15 +3639,20 @@ private int Enrol(int oFamilyId, int oInsureeId, int oPolicyId, int oPremiumId,
...
                         // Policy + premium
-                        JSONArray tempPoliciesArray = new JSONArray();
 
                         for (int j = 0; j < policiesArray.length(); j++) {
-                            tempPoliciesArray = policiesArray;
-                            JSONObject premiumObj = premiumsArray.getJSONObject(j);
-                            tempPoliciesArray.getJSONObject(j).put("premium", premiumObj);
+                            JSONArray policyPremiums = new JSONArray();
+                            String policyId = policiesArray.getJSONObject(j).getString("PolicyId");
+                            for(int k = 0; k < premiumsArray.length(); k++) {
```
Comment (kmadej) [inline app/src/main/java/org/openimis/imispolicies/ClientAndroidInterface.java:3646]: whitespace after "for"

**Review by kmadej — CHANGES_REQUESTED** [review]

**Comment by malinowskikam** [comment]
@kmadej I fixed the formatting. Can you review?

File: app/src/main/java/org/openimis/imispolicies/ClientAndroidInterface.java (hunk 3 above)
Comment (kmadej) [reply app/src/main/java/org/openimis/imispolicies/ClientAndroidInterface.java]: @malinowskikam you forgot about moving PolicyID string to constant

Comment (malinowskikam) [reply app/src/main/java/org/openimis/imispolicies/ClientAndroidInterface.java]: @jmadej I forgot to mention this, but every json object access in this project is done with string literals, and fixing this in only one place would just add into the confusion. (for instance this two literals actually mean different fields, and should be in 2 separate constants (one for policy schema and one for premium schema)

Comment (kmadej) [reply app/src/main/java/org/openimis/imispolicies/ClientAndroidInterface.java]: I see, thanks for explanation

**Review by kmadej — APPROVED** [review]

**Review by dragos-dobre — APPROVED** [review]

**Pull request merged**